packages/queued-request-controller/src/QueuedRequestController.ts

```diff
@@ -117,34 +151,77 @@ export class QueuedRequestController extends BaseController<
   }
 
   /**
-   * Switch the current globally selected network if necessary for processing the given
-   * request.
+   * Process the next batch of requests.
+   *
+   * This will trigger the next batch of requests with matching origins to be processed. Each
+   * request in the batch is dequeued one at a time, in chronological order, but they all get
+   * processed in parallel.
+   *
+   * This should be called after a batch of requests has finished processing, if the queue is non-
+   * empty.
+   */
+  async #processNextBatch() {
+    const firstRequest = this.#requestQueue.shift() as QueuedRequest;
+    this.#originOfCurrentBatch = firstRequest.origin;
+    const batch = [firstRequest.processRequest];
+    while (this.#requestQueue[0]?.origin === this.#originOfCurrentBatch) {
+      const nextEntry = this.#requestQueue.shift() as QueuedRequest;
+      batch.push(nextEntry.processRequest);
+    }
+
+    // If globally selected network is different from origin selected network,
+    // switch network before processing batch
+    let networkSwitchError: unknown;
+    try {
+      await this.#switchNetworkIfNecessary();
+    } catch (error: unknown) {
+      networkSwitchError = error;
+    }
+
+    batch.map(async (processRequest) => {
+      // These promises are handled as the return value of `#enqueueRequest`
+      // We don't need to handle them here
+      // eslint-disable-next-line @typescript-eslint/no-floating-promises
+      processRequest(networkSwitchError);
+    });
```

Comment:
does a forEach and dropping the async make sense here?
```suggestion
    batch.forEach((processRequest) => {
      // These promises are handled as the return value of `#enqueueRequest`
      // We don't need to handle them here
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      processRequest(networkSwitchError);
    });
```

Reply:
yeah this should be functionally equivalent I believe

Reply:
Good point! Simplified here: ff45f2aa

I went with a loop rather than a `forEach`, so that we could eliminate the extra function altogether.